x/escrow/keeper/internal/msg_server.go

```diff
@@ -0,0 +1,195 @@
+		Authority:         req.Authority,
+		MaxMetadataLength: req.MaxMetadataLength,
+	}); err != nil {
+		return nil, escrowv1alpha1.ErrInvariantBroken.Wrap(err.Error())
+	}
+
+	return &escrowv1alpha1.MsgUpdateParamsResponse{}, nil
+}
+
+func (s msgServer) CreateAgent(ctx context.Context, req *escrowv1alpha1.MsgCreateAgent) (*escrowv1alpha1.MsgCreateAgentResponse, error) {
+	if req.Creator == "" {
+		return nil, escrowv1alpha1.ErrUnimplemented.Wrap("nil creator")
+	}
+
+	creator, err := s.keeper.addressStringToBytes(req.Creator)
+	if err != nil {
+		return nil, errors.Wrap(err, "creator")
+	}
+
+	agent, err := s.keeper.CreateAgent(ctx, creator)
+	if err != nil {
+		return nil, err
+	}
+
+	agentStr, err := s.keeper.addressBytesToString(agent)
+	if err != nil {
+		return nil, errors.Wrap(escrowv1alpha1.ErrInvariantBroken.Wrap(err.Error()), "agent")
+	}
+
+	if err := sdk.UnwrapSDKContext(ctx).EventManager().EmitTypedEvent(&escrowv1alpha1.EventCreateAgent{
+		Agent:   agentStr,
+		Creator: req.Creator,
+	}); err != nil {
+		return nil, escrowv1alpha1.ErrInvariantBroken.Wrap(err.Error())
+	}
+
+	return &escrowv1alpha1.MsgCreateAgentResponse{Agent: agentStr}, nil
+}
+
+func (s msgServer) SubmitProposal(ctx context.Context, req *escrowv1alpha1.MsgSubmitProposal) (*escrowv1alpha1.MsgSubmitProposalResponse, error) {
+	if req.Proposer == "" {
+		return nil, escrowv1alpha1.ErrUnimplemented.Wrap("nil proposer")
+	}
+
+	if req.Agent == "" {
+		return nil, escrowv1alpha1.ErrUnimplemented.Wrap("nil agent")
+	}
+
+	if req.PreActions == nil {
+		return nil, escrowv1alpha1.ErrUnimplemented.Wrap("nil pre_actions")
+	}
+
+	if req.PostActions == nil {
+		return nil, escrowv1alpha1.ErrUnimplemented.Wrap("nil post_actions")
+	}
+
+	if req.Metadata == "" {
+		return nil, escrowv1alpha1.ErrUnimplemented.Wrap("nil metadata")
+	}
+
+	proposer, err := s.keeper.addressStringToBytes(req.Proposer)
+	if err != nil {
+		return nil, errors.Wrap(err, "proposer")
+	}
+
+	agent, err := s.keeper.addressStringToBytes(req.Agent)
+	if err != nil {
+		return nil, errors.Wrap(err, "agent")
+	}
+
+	signers := []sdk.AccAddress{proposer, agent}
+
+	if err := s.keeper.validateActions(req.PreActions, signers); err != nil {
+		return nil, errors.Wrap(err, "pre_actions")
+	}
+
+	if err := s.keeper.validateActions(req.PostActions, signers); err != nil {
+		return nil, errors.Wrap(err, "post_actions")
+	}
+
+	if err := s.keeper.SubmitProposal(ctx, proposer, agent, req.PreActions, req.PostActions, req.Metadata); err != nil {
+		return nil, err
+	}
+
+	if err := sdk.UnwrapSDKContext(ctx).EventManager().EmitTypedEvent(&escrowv1alpha1.EventSubmitProposal{
+		Proposer:    req.Proposer,
+		Agent:       req.Agent,
+		PreActions:  req.PreActions,
+		PostActions: req.PostActions,
+		Metadata:    req.Metadata,
+	}); err != nil {
+		return nil, escrowv1alpha1.ErrInvariantBroken.Wrap(err.Error())
+	}
+
+	return &escrowv1alpha1.MsgSubmitProposalResponse{}, nil
```

Comment:
This method of message server performs stateless verification, and `SubmitProposal()` does stateful verification. Hence unauthorized proposals cannot be submitted. The only concern would be submission of invalid proposals, but by design, there would be no other way but executing it. In that case, it's not obvious whether the reason of fault is stateful or not. So it would be almost impossible to prevent the submission.
It would be good to add periodic pruning logic to wipe out lingering invalid proposals.